### PR TITLE
feat(core): refresh interactable schemas after configuration changes

### PR DESCRIPTION
**Pull request by Kinfe123**
## Problem

Changing an interactable's `stateSchema` after mount leaves the registered definition on the previous schema. The model-facing update tool can therefore keep validating against stale state after a dynamically configured UI change.

On current main, rerendering either the current or legacy hook with a changed schema produced one registration instead of two. Multiple anchors made the stale model-facing schema persist even after re-registration.

## Root cause

Both hooks copied the latest schema into a ref, but their registration effects did not depend on the schema. Registration snapshots the supplied definition, so updating the ref alone cannot refresh it. The current registry also cached partial schemas only by interactable id, so a surviving anchor could retain the previous tool parameters.

## Change

A shared hook safely normalizes each supplied schema to JSON Schema, memoizes conversion by schema identity, and stabilizes structurally equivalent results. Current and legacy interactables now re-register when the normalized schema actually changes, including Standard Schema inputs, without churning when an equivalent schema is rebuilt inline. Unsupported schemas keep the existing permissive registry fallback instead of throwing during render.

The current registry tracks the schema source for each cached partial schema and rebuilds model-facing parameters when that source changes, including while another anchor remains registered. The original schema object remains registered so its validation behavior is preserved.

`initialState` keeps its existing ref behavior and does not trigger re-registration, so changing defaults cannot reset live state.

## Verification

- focused interactable suites: 63 tests passed
- full `@assistant-ui/core` suite: 1,843 tests passed
- `aui-build` for `@assistant-ui/core`
- focused Oxlint and Oxfmt checks
- bundle-size budget check

The changed-schema regressions fail on `origin/main` because each hook registers only once. Added coverage also proves equivalent rebuilt schemas remain stable, unsupported Standard Schemas do not fail render, stable schemas convert once, and multi-anchor model-context parameters refresh.

## Public surface

`@assistant-ui/core` receives a patch changeset. No exports or API signatures change.

### CODE REVIEW

File: packages/core/src/react/utils/useJSONEqual.ts [hunk 1]
```diff
@@ -1,6 +1,14 @@
 import { useRef } from "react";
 import { isJSONValueEqual } from "../../utils/json/is-json-equal";
 
+export function useJSONEqualValue<T>(value: T): T {
+  const prev = useRef(value);
+  if (prev.current !== value && !isJSONValueEqual(prev.current, value)) {
+    prev.current = value;
+  }
+  return prev.current;
+}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/utils/useJSONEqual.ts]: Fixed in bd1a6df89. The dependency now compares `toJSONSchema(stateSchema)`, so both JSON Schema and Standard Schema inputs use the normalized model-facing shape. `useJSONEqualValue` is constrained to `ReadonlyJSONValue`, and the original schema object is still passed to registration to preserve its validation behavior. Equivalent rebuilt Standard Schemas no longer cause registration churn.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 2]
```diff
@@ -138,6 +136,8 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
     id,
     name,
     config.description,
+    stateSchema,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts]: Agreed and fixed in bd1a6df89. `initialState` is back to its previous ref-only behavior and is no longer an effect dependency. Only a normalized schema change triggers re-registration, so changing defaults cannot introduce the nullish-state reset edge case.

File: packages/core/src/react/model-context/useInteractable.test.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    const schemaA = {
+      type: "object" as const,
+      properties: { first: { type: "string" } },
+    };
+    const schemaB = {
+      type: "object" as const,
+      properties: { second: { type: "number" } },
+    };
+    const initialA = { first: "one" };
+    const initialB = { second: 2 };
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.test.tsx:53]: Added the requested Standard Schema regression coverage in bd1a6df89 for both current and legacy hooks. Each test rebuilds an equivalent schema and confirms registration stays at one call, then changes the generated JSON Schema and confirms registration refreshes.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 4]
```diff
@@ -126,8 +124,8 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
       id,
       name,
       description: config.description,
-      stateSchema: stateSchemaRef.current,
-      initialState: initialStateRef.current,
+      stateSchema,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts]: Fixed in 37bcabad3. The current registry now records the source schema for each cached partial schema and recomputes the cache when that schema changes, even if another anchor keeps the id registered. A real registry test now keeps a second anchor mounted and asserts that getModelContext() exposes only the replacement schema's update-tool parameters.

File: packages/core/src/react/model-context/useInteractable.test.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+import { unstable_useInteractable } from "./useInteractable";
+
+afterEach(() => {
+  cleanup();
+  vi.clearAllMocks();
+});
+
+describe("unstable_useInteractable", () => {
+  it("refreshes the registration when its schema changes", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.test.tsx]: Addressed in 37bcabad3. initialState remains ref-only and an initialState-only rerender is covered without re-registration. The cache edge now has a separate test against the real Interactables resource and its getModelContext() parameters instead of relying on the mocked hook registration.

File: packages/core/src/react/utils/useJSONEqual.ts [hunk 6]
```diff
@@ -1,6 +1,14 @@
 import { useRef } from "react";
 import { isJSONValueEqual } from "../../utils/json/is-json-equal";
 
+export function useJSONEqualValue<T>(value: T): T {
+  const prev = useRef(value);
+  if (prev.current !== value && !isJSONValueEqual(prev.current, value)) {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/utils/useJSONEqual.ts:7]: Fixed in 37bcabad3. Both hooks now use a shared useJSONSchemaDependency helper that normalizes Standard Schemas before deep stabilization. Equivalent rebuilt schemas retain the same dependency, while a changed normalized schema refreshes registration.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 7]
```diff
@@ -116,6 +118,9 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
 
   const stateSchemaRef = useRef(config.stateSchema);
   stateSchemaRef.current = config.stateSchema;
+  const normalizedStateSchema = useJSONEqualValue(
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts]: Fixed in 37bcabad3. Conversion is now shared, memoized by schema identity, wrapped in try/catch, and runtime-validated with isJSONValue before deep comparison. Unsupported schemas use a stable null dependency and still reach the registry's existing degraded fallback. The core build and all 1,843 core tests pass.

File: packages/core/src/react/interactables-legacy/useAssistantInteractable.ts [hunk 8]
```diff
@@ -38,6 +41,9 @@ export const useAssistantInteractable = (
 
   const stateSchemaRef = useRef(config.stateSchema);
   stateSchemaRef.current = config.stateSchema;
+  const normalizedStateSchema = useJSONEqualValue(
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/interactables-legacy/useAssistantInteractable.ts]: Fixed in 37bcabad3 with the same shared helper used by the current hook, so there is no duplicated conversion logic. I kept the legacy path aligned because it remains a shipped public hook until removal; its focused regressions cover equivalent rebuilt and unsupported Standard Schemas.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 9]
```diff
@@ -116,6 +117,9 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
...
+    toJSONSchema(config.stateSchema),
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts]: Fixed in 37bcabad3. Schema normalization is now non-throwing and memoized in a shared hook. A Standard Schema without a JSON converter produces a stable null dependency, registers normally, and is covered by focused tests for both current and legacy hooks.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 10]
```diff
@@ -116,6 +118,9 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
 
   const stateSchemaRef = useRef(config.stateSchema);
   stateSchemaRef.current = config.stateSchema;
+  const normalizedStateSchema = useJSONEqualValue(
+    toJSONSchema(config.stateSchema) as ReadonlyJSONValue,
+  );
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts]: Fixed in 37bcabad3. The shared helper memoizes conversion for stable schema references, catches converter failures, validates successful output as JSON, and stabilizes only validated JSON or null. This preserves the registry fallback and avoids per-render registration churn; focused and full core tests pass.

File: packages/core/src/react/utils/useJSONEqual.ts [hunk 11]
```diff
@@ -1,6 +1,15 @@
...
 import { isJSONValueEqual } from "../../utils/json/is-json-equal";
 
+export function useJSONEqualValue<T extends ReadonlyJSONValue>(value: T): T {
+  const prev = useRef(value);
+  if (prev.current !== value && !isJSONValueEqual(prev.current, value)) {
+    prev.current = value;
+  }
+  return prev.current;
+}
+
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/utils/useJSONEqual.ts:12]: Fixed in 37bcabad3 without a runtime cast. useJSONSchemaDependency checks the converter output with isJSONValue and passes useJSONEqualValue only validated JSON or the stable null fallback. Unknown non-JSON values therefore cannot reach the memo helper, and unsupported rebuilt schemas are covered by regression tests.

File: packages/core/src/react/model-context/useInteractable.ts [hunk 12]
```diff
@@ -138,6 +143,7 @@ const useInteractable = <TSchema extends Unstable_InteractableStateSchema>(
     id,
     name,
     config.description,
+    normalizedStateSchema,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.ts:143]: Fixed in 37bcabad3. partialSchemaSourceRef tracks the raw schema behind each cached partial schema, and registration invalidates/rebuilds the cache when that source changes regardless of remaining anchors. The new registry test keeps another anchor alive and asserts the actual update_note parameters switch from the old property to the new one.

File: packages/core/src/react/model-context/useInteractable.test.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/model-context/useInteractable.test.tsx]: @rupic-app The isolated coverage is in the new `converts a referentially stable schema only once` test: it keeps the same `stateSchema` object, rerenders with only a new `initialState`, and asserts both one conversion and one registration. The earlier changed-schema test remains separate.

File: packages/core/src/react/interactables-legacy/useAssistantInteractable.ts [hunk 13]
```diff
@@ -38,6 +41,9 @@ export const useAssistantInteractable = (
...
+    toJSONSchema(config.stateSchema) as ReadonlyJSONValue,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/interactables-legacy/useAssistantInteractable.ts]: Fixed in 37bcabad3. Both hooks now call a shared helper that memoizes by schema identity, catches conversion failures, and uses a stable null fallback. The raw opaque schema is intentionally not used as the dependency because rebuilt unsupported schema instances cannot be compared and would re-register on every render. Focused tests cover successful registration and no churn for unsupported rebuilt Standard Schemas.

File: packages/core/src/react/interactables-legacy/useAssistantInteractable.ts [hunk 14]
```diff
@@ -50,7 +56,14 @@ export const useAssistantInteractable = (
       initialState: initialStateRef.current,
       selected: config.selected,
     });
-  }, [aui, id, name, config.description, config.selected]);
+  }, [
+    aui,
+    id,
+    name,
+    config.description,
+    normalizedStateSchema,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/interactables-legacy/useAssistantInteractable.ts:60]: The applicable current-client cache issue is fixed in 37bcabad3 by tracking each cache entry's source schema and rebuilding it when that source changes. The legacy registry does not have the guarded cache-hit path: it recomputes and replaces `partialSchemaCacheRef` unconditionally on every registration, so the reported stale-cache mechanism does not apply there.

File: packages/core/src/react/interactables-legacy/useAssistantInteractable.test.tsx [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    );
+    await waitFor(() => expect(mocks.register).toHaveBeenCalledTimes(1));
+
+    hook.rerender({ stateSchema: createSchema("value") });
+    expect(mocks.register).toHaveBeenCalledTimes(1);
+
+    const changedSchema = createSchema("other");
+    hook.rerender({ stateSchema: changedSchema });
+    await waitFor(() => expect(mocks.register).toHaveBeenCalledTimes(2));
+    expect(mocks.register).toHaveBeenLastCalledWith(
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/interactables-legacy/useAssistantInteractable.test.tsx:103]: Added the requested real multi-anchor regression in 37bcabad3 at `packages/core/src/react/client/Interactables.test.ts`. It keeps a second anchor mounted and asserts `getModelContext().tools.update_note.parameters` contains the replacement property and not the old one. That contract belongs to the current registry; the legacy registry has no registration-count cache and recomputes on every register.

File: packages/core/src/react/interactables-legacy/useAssistantInteractable.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/interactables-legacy/useAssistantInteractable.ts]: Fixed in 37bcabad3. The shared `useJSONSchemaDependency` helper wraps conversion in `useMemo(..., [schema])`, and the focused current-hook test rerenders with the same schema identity and verifies its converter runs only once.

File: packages/core/src/react/utils/useJSONSchemaDependency.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+export function useJSONSchemaDependency(
+  schema: Unstable_InteractableStateSchema,
+): ReadonlyJSONValue {
+  const normalizedSchema = useMemo(() => {
+    try {
+      const jsonSchema = toJSONSchema(schema);
+      return isJSONValue(jsonSchema) ? jsonSchema : null;
+    } catch {
+      return null;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/utils/useJSONSchemaDependency.ts:17]: Recorded in e267b2434. The helper now explains why opaque schemas deliberately share one stable fallback: their semantic shape cannot be compared, and keying the dependency by rebuilt object identity would churn registration on every render. The current path remains correct and the legacy limitation stays unchanged ahead of its scheduled removal.

File: packages/core/src/react/client/Interactables.test.ts [hunk 17]
```diff
@@ -321,6 +326,41 @@ describe("Interactables registration", () => {
...
+    };
+    root = mount({ threadMessages: [createCall("n1")] });
+    const first = root.getValue().register(reg("n1", { stateSchema: schemaA }));
+    const second = root
+      .getValue()
+      .register(reg("n1", { stateSchema: schemaA }));
+    await flushMicrotasks();
+
+    first();
+    root.getValue().register(reg("n1", { stateSchema: schemaB }));
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/client/Interactables.test.ts]: Fixed in 6a0aa8852. The replacement registration cleanup is now retained and called after the remaining original anchor, so the multi-anchor regression leaves the registry at a zero refcount. The focused registry suite passes all 29 tests.
